provider/pkg/await/daemonset.go

```diff
@@ -0,0 +1,356 @@
+	if dsa.config.timeout != nil {
+		timeout = *dsa.config.timeout
+	}
+	ctx, cancel := context.WithCancelCause(dsa.config.ctx)
+	defer cancel(context.Canceled)
+	go func() {
+		dsa.config.Clock().Sleep(timeout)
+		cancel(context.DeadlineExceeded)
+	}()
+
+	stopper := make(chan struct{})
+	defer close(stopper)
+
+	informerFactory := informers.NewInformerFactory(
+		dsa.config.clientSet,
+		informers.WithNamespaceOrDefault(dsa.config.currentOutputs.GetNamespace()),
+	)
+	informerFactory.Start(stopper)
+
+	dsEvents := make(chan watch.Event)
+	dsInformer, err := informers.New(
+		informerFactory,
+		informers.ForGVR(v1.SchemeGroupVersion.WithResource("daemonsets")),
+		informers.WithEventChannel(dsEvents),
+	)
+	if err != nil {
+		return err
+	}
+	go dsInformer.Informer().Run(stopper)
+
+	podEvents := make(chan watch.Event)
+	podInformer, err := informers.New(
+		informerFactory,
+		informers.ForPods(),
+		informers.WithEventChannel(podEvents),
+	)
+	if err != nil {
+		return err
+	}
+	go podInformer.Informer().Run(stopper)
+
+	podAggregator := NewPodAggregator(ResourceIDFromUnstructured(dsa.ds), podInformer.Lister())
+	podAggregator.Start(podEvents)
+	defer podAggregator.Stop()
+
+	for {
+		if done() {
+			return nil
+		}
+		select {
+		case <-ctx.Done():
+			return wait.ErrorInterrupted(nil)
+		case event := <-dsEvents:
+			dsa.processDaemonSetEvent(event)
+		case messages := <-podAggregator.ResultChan():
+			dsa.processPodMessages(messages)
+		}
+	}
+}
+
+// onDeleteRolloutStatus returns true when a DaemonSet with the OnDelete
+// rollout strategy has been fully rolled out. This strategy requires pods to
+// be deleted out-of-band.
+//
+// Success occurs when:
+//  1. generation <= observedGeneration
+//  2. numberAvailable == desiredNumberScheduled, and
+//  3. numberMisscheduled == 0.
+func (dsa *dsAwaiter) onDeleteRolloutStatus() (string, bool, error) {
+	ds := dsa.ds.Object
+
+	generation, _, _ := unstructured.NestedInt64(ds, "generation")
+	observedGen, _, _ := unstructured.NestedInt64(ds, "status", "observedGeneration")
+
+	if generation > observedGen {
+		return "Waiting for controller to reconcile...", false, nil
+	}
+
+	dns, _, _ := unstructured.NestedInt64(ds, "status", "desiredNumberScheduled")
+	na, _, _ := unstructured.NestedInt64(ds, "status", "numberAvailable")
+	nms, _, _ := unstructured.NestedInt64(ds, "status", "numberMisscheduled")
+
+	msg := "OnDelete rollout, waiting for old pods to be deleted..."
+	done := (na == dns) && (nms == 0)
+	if done {
+		msg = "OnDelete rollout complete"
+	}
+	return msg, done, nil
+}
+
+// rolloutComplete checks whether we've succeeded, and logs the result as a
+// status message to the provider.
+func (dsa *dsAwaiter) rolloutComplete() bool {
+	dssv := polymorphichelpers.DaemonSetStatusViewer{}
+
+	msg, done, err := dssv.Status(dsa.ds, 0)
+	if done && err != nil {
+		// If done=true and err is non-nil then this is an OnDelete rollout.
+		msg, done, err = dsa.onDeleteRolloutStatus()
```

Comment:
This is a pretty awkward handoff, and given how little code is in `DaemonSetStatusViewer`, maybe we shouldn't use it.

Reply:
I switched this over to use kstatus since it's essentially the same logic and doesn't distinguish between Rolling or OnDelete rollouts.